File: android/app/src/main/java/id/omi/absensicafe/domain/Pin.kt
package id.omi.absensicafe.domain

import id.omi.absensicafe.data.model.Employee
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec

/**
 * PIN pribadi karyawan.
 *
 * PIN hanya empat angka, jadi hash cepat seperti SHA-256 polos bisa ditebak
 * habis dalam hitungan detik kalau isinya bocor. Karena itu dipakai
 * PBKDF2-HMAC-SHA256 dengan garam per karyawan: menebak 10.000 kemungkinan
 * jadi butuh waktu, sementara satu kali pemeriksaan di tablet tetap sekejap.
 *
 * Algoritme dan jumlah putarannya sama persis dengan yang dipakai web admin
 * lewat WebCrypto, sehingga PIN yang dibuat di PC bisa diperiksa di tablet.
 */
object Pin {

    const val LENGTH = 4
    const val ITERATIONS = 120_000
    private const val KEY_BITS = 256

    private val random = SecureRandom()

    fun isValidFormat(pin: String): Boolean =
        pin.length == LENGTH && pin.all { it.isDigit() }

    fun newSalt(): String {
        val bytes = ByteArray(16)
        random.nextBytes(bytes)
        return bytes.toHex()
    }

    fun hash(pin: String, saltHex: String, iterations: Int = ITERATIONS): String {
        val spec = PBEKeySpec(pin.toCharArray(), saltHex.fromHex(), iterations, KEY_BITS)
        val factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256")
        return factory.generateSecret(spec).encoded.toHex()
    }

    /** Perbandingan waktu tetap, supaya lama proses tidak membocorkan tebakan. */
    fun verify(pin: String, employee: Employee): Boolean {
        if (!employee.hasPin || !isValidFormat(pin)) return false
        val iterations = if (employee.pinIterations > 0) employee.pinIterations else ITERATIONS
        val calon = hash(pin, employee.pinSalt, iterations)
        return MessageDigest.isEqual(
            calon.toByteArray(Charsets.US_ASCII),
            employee.pinHash.toByteArray(Charsets.US_ASCII)
        )
    }

    private fun ByteArray.toHex(): String =
        joinToString("") { "%02x".format(it) }

    private fun String.fromHex(): ByteArray =
        ByteArray(length / 2) { i -> substring(i * 2, i * 2 + 2).toInt(16).toByte() }
}

File: web-admin/src/lib/pin.ts
/**
 * PIN pribadi karyawan.
 *
 * Harus menghasilkan hash yang identik dengan `domain/Pin.kt` di aplikasi
 * Android, karena PIN dibuat di sini lalu diperiksa di tablet: sama-sama
 * PBKDF2-HMAC-SHA256, 120.000 putaran, kunci 256 bit, garam 16 byte acak.
 *
 * PIN hanya empat angka. Untuk digit ASCII, penyandian UTF-8 di WebCrypto dan
 * penyandian char[] di Java menghasilkan byte yang sama, jadi kedua sisi tidak
 * perlu penyesuaian apa pun.
 */

export const PIN_LENGTH = 4;
export const PIN_ITERATIONS = 120_000;
const KEY_BITS = 256;

export function isValidPinFormat(pin: string): boolean {
  return pin.length === PIN_LENGTH && /^\d+$/.test(pin);
}

export function newSalt(): string {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  return toHex(bytes);
}

export async function hashPin(
  pin: string,
  saltHex: string,
  iterations: number = PIN_ITERATIONS,
): Promise<string> {
  const key = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(pin),
    "PBKDF2",
    false,
    ["deriveBits"],
  );
  const bits = await crypto.subtle.deriveBits(
    {
      name: "PBKDF2",
      salt: fromHex(saltHex) as BufferSource,
      iterations,
      hash: "SHA-256",
    },
    key,
    KEY_BITS,
  );
  return toHex(new Uint8Array(bits));
}

/** Membuat garam baru sekaligus hash-nya, untuk disimpan di dokumen karyawan. */
export async function makePinFields(pin: string): Promise<{
  pinSalt: string;
  pinHash: string;
  pinIterations: number;
}> {
  const pinSalt = newSalt();
  const pinHash = await hashPin(pin, pinSalt);
  return { pinSalt, pinHash, pinIterations: PIN_ITERATIONS };
}

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

function fromHex(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}
